File: src/901-1000/998_brute.py
import numpy as np
from scipy.optimize import minimize_scalar
import math
import time

def get_min_bounding_square(a, b, c):
    """
    Calculates exact minimum bounding square side using Brent's optimization method.
    """
    x_A = (a**2 + c**2 - b**2) / (2 * a)
    y_A_sq = c**2 - x_A**2
    
    if y_A_sq <= 0:
        return 0.0 # Degenerate triangle
        
    y_A = math.sqrt(y_A_sq)
    
    Vx = np.array([0.0, a, x_A])
    Vy = np.array([0.0, 0.0, y_A])
    
    def bounding_square_size(theta):
        cos_t, sin_t = math.cos(theta), math.sin(theta)
        
        # Rotate coordinates
        X = Vx * cos_t - Vy * sin_t
        Y = Vx * sin_t + Vy * cos_t
        
        W = np.max(X) - np.min(X)
        H = np.max(Y) - np.min(Y)
        return max(W, H)

    # The max(W, H) function can have a sharp peak where W = H.
    # To avoid getting stuck in a false local minimum, we evaluate two chunks of the 90-degree quadrant.
    best_min = float('inf')
    for start_angle in [0, np.pi/4]:
        res = minimize_scalar(
            bounding_square_size, 
            bounds=(start_angle, start_angle + np.pi/4), 
            method='bounded',
            options={'xatol': 1e-9}
        )
        if res.fun < best_min:
            best_min = res.fun
            
    return best_min

def is_integer_bounding_square(a, b, c, tol=1e-5):
    min_sq = get_min_bounding_square(a, b, c)
    nearest_int = round(min_sq)
    
    if abs(min_sq - nearest_int) < tol:
        return True, nearest_int, min_sq
    return False, nearest_int, min_sq

def compute_T(n):
    print(f"--- Computing T({n}) ---")
    start_time = time.time()
    
    total_perimeter = 0
    max_side = int(n * math.sqrt(2)) + 1
    
    print(f"{'Triangle (a, b, c)':<20} | {'Square Size':<12} | {'Perimeter'}")
    print("-" * 55)
    
    for a in range(1, max_side + 1):
        for b in range(1, a + 1):
            for c in range(a - b + 1, b + 1):
                
                is_int, sq_size, _ = is_integer_bounding_square(a, b, c)
                
                if is_int and sq_size <= n:
                    p = a + b + c
                    total_perimeter += p
                    print(f"({a:>2}, {b:>2}, {c:>2})          | {sq_size:>12} | {p:>4}")

    elapsed = time.time() - start_time
    print("-" * 55)
    print(f"T({n}) = {total_perimeter}")
    print(f"Time taken: {elapsed:.3f} seconds\n")

if __name__ == "__main__":
    is_int, val, raw = is_integer_bounding_square(5, 4, 3)
    print(f"Triangle (3, 4, 5): Minimum Square = {raw:.5f} | Is Integer? {is_int}")
    
    is_int, val, raw = is_integer_bounding_square(13, 13, 10)
    print(f"Triangle (10, 13, 13): Minimum Square = {raw:.5f} | Is Integer? {is_int}\n")
    
    compute_T(40)

File: src/901-1000/998_brute.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <iomanip>
#include <chrono>
#include <algorithm>

// Define a bulletproof PI
const double PI = std::acos(-1.0);

const int NUM_ANGLES = 360000;
std::vector<double> cos_t(NUM_ANGLES);
std::vector<double> sin_t(NUM_ANGLES);

void precompute_angles() {
    for (int i = 0; i < NUM_ANGLES; ++i) {
        // Swap M_PI for our defined PI
        double theta = (PI / 2.0) * i / (NUM_ANGLES - 1.0);
        cos_t[i] = std::cos(theta);
        sin_t[i] = std::sin(theta);
    }
}

double get_min_bounding_square(double a, double b, double c) {
    double x_A = (a*a + c*c - b*b) / (2.0 * a);
    double y_A_sq = c*c - x_A*x_A;
    
    if (y_A_sq <= 0.0) return 0.0;
    
    double y_A = std::sqrt(y_A_sq);
    
    double vx[3] = {0.0, a, x_A};
    double vy[3] = {0.0, 0.0, y_A};
    
    double min_sq = 1e18; // Start with a massive float
    
    // Dense angle sweep
    for (int i = 0; i < NUM_ANGLES; ++i) {
        double ct = cos_t[i];
        double st = sin_t[i];
        
        double min_x = 1e18, max_x = -1e18;
        double min_y = 1e18, max_y = -1e18;
        
        for (int j = 0; j < 3; ++j) {
            double px = vx[j] * ct - vy[j] * st;
            double py = vx[j] * st + vy[j] * ct;
            
            if (px < min_x) min_x = px;
            if (px > max_x) max_x = px;
            if (py < min_y) min_y = py;
            if (py > max_y) max_y = py;
        }
        
        double w = max_x - min_x;
        double h = max_y - min_y;
        double sq_size = std::max(w, h);
        
        if (sq_size < min_sq) {
            min_sq = sq_size;
        }
    }
    return min_sq;
}

void compute_T(int n) {
    std::cout << "--- Computing T(" << n << ") [C++ Multithreaded Sweep] ---\n";
    auto start_time = std::chrono::high_resolution_clock::now();
    
    long long total_perimeter = 0;
    int max_side = static_cast<int>(n * std::sqrt(2.0)) + 1;
    double tol = 1e-5;

    // Utilize all available cores
    #pragma omp parallel for reduction(+:total_perimeter) schedule(dynamic)
    for (int a = 1; a <= max_side; ++a) {
        for (int b = 1; b <= a; ++b) {
            for (int c = a - b + 1; c <= b; ++c) {
                
                double min_sq = get_min_bounding_square(a, b, c);
                if (min_sq == 0.0) continue;
                
                double nearest_int = std::round(min_sq);
                
                if (std::abs(min_sq - nearest_int) < tol && nearest_int <= n) {
                    long long p = a + b + c;
                    total_perimeter += p;
                    
                    // Critical section for clean I/O printing
                    #pragma omp critical
                    {
                        std::cout << "(" << std::setw(2) << a << ", " 
                                  << std::setw(2) << b << ", " 
                                  << std::setw(2) << c << ") | "
                                  << std::fixed << std::setprecision(6) << min_sq 
                                  << " -> " << static_cast<int>(nearest_int) 
                                  << " | P = " << p << "\n";
                    }
                }
            }
        }
    }

    auto end_time = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> elapsed = end_time - start_time;

    std::cout << "-------------------------------------------------------\n";
    std::cout << "T(" << n << ") = " << total_perimeter << "\n";
    std::cout << "Time taken: " << std::fixed << std::setprecision(3) << elapsed.count() << " seconds\n";
}

int main() {
    // Setup precomputed tables
    precompute_angles();
    
    // Explicit format setup to prevent e-notation
    std::cout << std::fixed << std::setprecision(6);
    
    // Test the specific edge case
    double test_val = get_min_bounding_square(19, 19, 17);
    std::cout << "Absolute pure minimum for (19, 19, 17) is: " << test_val << "\n\n";
    
    compute_T(40);
    return 0;
}